Fix `SegmentedButton` clipping when drawing segments

TahaTesser opened the pull request:
fixes [`SegmentedButton` doesn't clip properly when one of the segments has `ColorFiltered`](https://github.com/flutter/flutter/issues/144990)

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: SegmentedButton<int>(
            segments: const <ButtonSegment<int>>[
              ButtonSegment<int>(
                value: 0,
                label: ColorFiltered(
                  colorFilter:
                      ColorFilter.mode(Colors.amber, BlendMode.colorBurn),
                  child: Text('Option 1'),
                ),
              ),
              ButtonSegment<int>(
                value: 1,
                label: Text('Option 2'),
              ),
            ],
            onSelectionChanged: (Set<int> selected) {},
            selected: const <int>{0},
          ),
        ),
      ),
    );
  }
}
```

</details>

### Before

![before](https://github.com/flutter/flutter/assets/48603081/b402fc51-d575-4208-8a71-f798ef2b2bf5)


### After

![after](https://github.com/flutter/flutter/assets/48603081/77a5cac2-ecef-4381-a043-dbb5c91407ec)


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

TahaTesser committed:
Fix `SegmentedButton` clipping when drawing segments

QuncCccccc reviewed (commented):
Could you also help try other platforms? It seems the vertical line is still not clipped on mobile devices but correctly clipped on maxOS and web🧐 The splash effect when we press the button is clipped correctly with this change though:)
![Screenshot 2024-06-05 at 5 04 33 PM](https://github.com/flutter/flutter/assets/36861262/01975cf2-21cf-4da4-94f5-3bc6d73c1eac)

TahaTesser committed:
Fix divider offset

TahaTesser commented:
> Could you also help try other platforms? It seems the vertical line is still not clipped on mobile devices but correctly clipped on maxOS and web🧐 The splash effect when we press the button is clipped correctly with this change though:)

Thank you so much! I should've tested this on mobile. 

The reason divider is out the border because previous it was using `childRect` to clip the canvas and paint the child.  In my fix, I'm using border path to clip the canvas. The divider bug was due to it is still positing child rect, I fixed this by also updating divider offset and added a new test to catch this bug in the future.

| Preview | 
| --------------- | 
| <img src="https://github.com/flutter/flutter/assets/48603081/746b41d1-fb17-48f7-80df-25f595f0e192" height="650" />	|

QuncCccccc reviewed (approved):
> The divider bug was due to it is still using the childRect to position, I fixed this by updating divider offset and added a new test to catch this bug in the future.

Ah, this makes sense to me. Thanks for the explanation and the test! LGTM:)

TahaTesser commented:
@QuncCccccc 
This is failing Google Testing, it might be legit. Can you please take a look?

QuncCccccc commented:
> @QuncCccccc This is failing Google Testing, it might be legit. Can you please take a look?

Checking!

QuncCccccc commented:
Reason for revert: the newly-added unit test might cause the red tree
